File: src/main/java/io/github/support/terminal/platform/app/domains/core/source/dbs/models/DbConnectionDTO.java
package io.github.support.terminal.platform.app.domains.core.source.dbs.models;


import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import io.github.support.terminal.platform.app.domains.core.source.dbs.values.DbType;
import lombok.Data;
import lombok.EqualsAndHashCode;


@Data
@EqualsAndHashCode(of={"id"})
@JsonTypeInfo(
        use = JsonTypeInfo.Id.NAME,
        include = JsonTypeInfo.As.PROPERTY,
        property = "type",
        visible = true)
@JsonSubTypes({
        @JsonSubTypes.Type(value = PostgresDbConnectionDTO.class, name = DbType.POSTGRES),
        @JsonSubTypes.Type(value = OracleDbConnectionDTO.class, name = DbType.ORACLE),
        @JsonSubTypes.Type(value = MySqlDbConnectionDTO.class, name = DbType.MYSQL)
})
public abstract class DbConnectionDTO {
    private String id;
    private String name;
    private String type;
    private String host;
    private Integer port;
    private String user;
    private String password;
}

File: src/main/kotlin/io/github/support/terminal/application/domains/db_connection/rest/DbConnectionDTO.kt
package io.github.support.terminal.application.domains.db_connection.rest

import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonTypeInfo
import io.github.support.terminal.application.domains.db_connection.entity.DbConnectionType

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type", visible = true)
@JsonSubTypes(
        JsonSubTypes.Type(value = PostgresDbConnectionDTO::class, name = DbConnectionType.Constants.POSTGRES),
        JsonSubTypes.Type(value = OracleDbConnectionDTO::class, name = DbConnectionType.Constants.ORACLE),
        JsonSubTypes.Type(value = MySqlDbConnectionDTO::class, name = DbConnectionType.Constants.MYSQL))
abstract class DbConnectionDTO(
        val type: String
)

data class MySqlDbConnectionDTO (
        val id: String,
        val name: String,
        val host: String,
        val port: Int,
        val user: String,
        val password: String,
        val dbName: String
): DbConnectionDTO(DbConnectionType.Constants.MYSQL)

data class OracleDbConnectionDTO (
        val id: String,
        val name: String,
        val host: String,
        val port: Int,
        val user: String,
        val password: String,
        val sid: String
): DbConnectionDTO(DbConnectionType.Constants.ORACLE)

data class PostgresDbConnectionDTO (
        val id: String,
        val name: String,
        val host: String,
        val port: Int,
        val user: String,
        val password: String,
        val dbName: String
): DbConnectionDTO(DbConnectionType.Constants.POSTGRES)
